### PR TITLE
Fix and simplify the etcdv2 Syncer

### PR DESCRIPTION
**Pull request by fasaxc**
I found a potential race and a real bug in the etcdv2 Syncer.  This change reworks the Syncer to fix the issue and simplify the snapshot start processing.

The bug was that, after the watcher went out-of-sync, it wouldn't start a new snapshot until it received the next event.  Hence, if nothing was going on, you could be out-of-sync indefinitely.  The watcher now explicitly establishes the current etcd index by reading the Ready flag (which has the nice side effect of stalling the watcher until the Ready flag appears).

The snapshot initiation logic was based on a dubious triangle of channels and I couldn't convince myself that there wasn't a race lurking in there.  I've simplified that so that the watcher thread tells the merge thread that it needs a new snapshot; then the merge thread sends the message to the snapshot thread.  That way there can't be a race between merge thread being told a snapshot is starting and receiving the first update from the snapshot thread.

Along the way, I also cleaned up the channels to use real message types rather than a single struct for everything.  That makes the unpacking code much cleaner.

### CODE REVIEW

**Comment by fasaxc** [comment]
This has passed the OpenStack CI.  I think we should merge it soon so that it has some time to bake before the GA.

**Comment by fasaxc** [comment]
@robbrockbank did you say you'd take a crack at reviewing this?

File: lib/backend/model/ippool.go [hunk 1]
```diff
@@ -28,7 +28,7 @@ import (
 
 var (
 	matchIPPool = regexp.MustCompile("^/?calico/v1/ipam/v./pool/([^/]+)$")
-	typeIPPool = reflect.TypeOf(IPPool{})
+	typeIPPool  = reflect.TypeOf(IPPool{})
```
Comment (caseydavenport) [inline lib/backend/model/ippool.go]: extra space here

Comment (fasaxc) [inline lib/backend/model/ippool.go]: That's a `go fmt` cleanup so I'll leave as-is.

**Pull request merged**